### PR TITLE
feat(suite-native): Graph package

### LINKED ISSUE #5712: suite-native: Graphs
**Design:** [Figma](https://www.figma.com/file/Z6AGVUmKQzLNtDozFamW7f/s2-Mobile?node-id=1246%3A10364)

**Spec:** 
1. Use library https://github.com/margelo/react-native-graph
2. This would be nice to have in standalone package that we will only use at Dashboard screen
3. For this we can go with default graph visual from library maybe just tune colors little bit and we will fine tune appearnce later
4. Legend switch under graph should part of package too.

### PR DESCRIPTION
**Pull request by juriczech**
https://tppr.me/k8vWs

Closes https://github.com/trezor/trezor-suite/issues/5712

@dahaca Fixes some comments from this PR https://github.com/trezor/trezor-suite/pull/5818

### CODE REVIEW

**Comment by Nodonisko** [comment]
This package is safe, it's quite new and first of it's kind build on top of Skia, that's reason why it doesn't have many downloads. Also author is guy that I know from conferences.

Other alternatives looks very shitty and not to  much interactive and customizable. Since this it's made using Skia, it has great potential, but maybe we will have to contribute to the package.

File: CODEOWNERS [hunk 1]
```diff
@@ -46,7 +46,7 @@
 
 /packages/suite-desktop-api @szymonlesisz @marekrjpolak
 
-/suite-native @Nodonisko @bouchja1 @juriczech
+/suite-native @Nodonisko @bouchja1 @juriczech @dahaca
```
Comment (juriczech) [inline CODEOWNERS:49]: I added @dahaca since he's meant to review the UI so he can see our PR's in time and review the changes against design.

File: suite-native/graph/src/reducers/types.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+export type TimeFrameValues = 'hour' | 'day' | 'week' | 'month' | 'year' | 'all';
```
Comment (Nodonisko) [inline suite-native/graph/src/reducers/types.ts]: why is this file in `reducers` folder?

File: suite-native/state/src/configureStore.ts [hunk 3]
```diff
@@ -1,11 +1,13 @@
...
+import { combineReducers, configureStore, Store } from '@reduxjs/toolkit';
 
 import { appSettingsReducer } from '@suite-native/module-settings';
+import { appGraphReducer } from '@suite-native/graph';
 
-export const store = configureStore({
-    reducer: {
-        appSettings: appSettingsReducer,
-    },
+const rootReducer = combineReducers({
```
Comment (Nodonisko) [inline suite-native/state/src/configureStore.ts]: Why these changes? What wrong with configure store?

File: suite-native/graph/src/reducers/appGraphReducer.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+import { TimeFrameValues } from './types';
+
+interface AppGraphState {
+    selectedTimeFrame: TimeFrameValues;
+}
+
+const initialState = {
+    selectedTimeFrame: 'day',
+} as AppGraphState;
```
Comment (Nodonisko) [inline suite-native/graph/src/reducers/appGraphReducer.ts]: Do use type casting, just assign type `const initialState: AppGraphState = ...`

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-native/graph/src/reducers/types.ts (hunk 2 above)
Comment (juriczech) [reply suite-native/graph/src/reducers/types.ts]: I wanted to have everything related to the logic of package in one folder. I copied `/reducers/packageNameReducer.ts` naming convention from settings, but maybe I could rename the folder to sth else. Any idea? Or I can keep this files in src folder of this pkg.

File: suite-native/state/src/configureStore.ts (hunk 3 above)
Comment (juriczech) [reply suite-native/state/src/configureStore.ts]: No particular reason. I was debugging and trying some stuff and decided to leave it like this as it's recommended in the docs but it's basically the same. Do you think it's unwanted? Nothing wrong with how it was as far as I'm concerned.

File: suite-native/graph/src/reducers/appGraphReducer.ts (hunk 4 above)
Comment (juriczech) [reply suite-native/graph/src/reducers/appGraphReducer.ts]: https://github.com/trezor/trezor-suite/pull/5838/commits/329ca57f9fe04f16cd52896651bcc2883ab5f681

File: suite-native/graph/src/reducers/types.ts (hunk 2 above)
Comment (bouchja1) [reply suite-native/graph/src/reducers/types.ts]: > Or I can keep this files in src folder of this pkg

I would go that way... `types.ts` in src.

File: suite-native/state/src/configureStore.ts (hunk 3 above)
Comment (bouchja1) [reply suite-native/state/src/configureStore.ts]: @juriczech Yes, it is possible to do it in this way but I am sure there is nothing like "recommended" for store config in the [documentation](https://redux-toolkit.js.org/usage/usage-with-typescript#getting-the-state-type). It's just about typing. I think we are trying to use the new stuff where redux-toolkit (I mean configureStore) calls combineReducers automatically. It is not necessary to do it directly.

I personally prefer the second way without combining reducers:

> Alternatively, if you choose to not create a rootReducer yourself and instead pass the slice reducers directly to configureStore(), you need to slightly modify the typing to correctly infer the root reducer

Comment (juriczech) [reply suite-native/state/src/configureStore.ts]: @bouchja1 I just read [this](https://tppr.me/z2S7C), where it's mentioned that this might me more understandable for others but imo, I don't see why:D I just thought I'll leave it like this without giving it a second thought. I'll change it back to `configureStore`.

Comment (juriczech) [reply suite-native/state/src/configureStore.ts]: Also, since this file's main purpose is to export store, shouldn't the name be `store.ts`?

Comment (juriczech) [reply suite-native/state/src/configureStore.ts]: Fixed. @Nodonisko

Comment (bouchja1) [reply suite-native/state/src/configureStore.ts]: > Also, since this file's main purpose is to export store, shouldn't the name be `store.ts`?

well, here we can say it is somehow "[recommended](https://redux-toolkit.js.org/tutorials/quick-start#create-a-redux-store)" :D

File: suite-native/graph/src/reducers/types.ts (hunk 2 above)
Comment (Nodonisko) [reply suite-native/graph/src/reducers/types.ts]: Definitely in root, no reason to keep it only in reducers folder when it's used also in components etc.

Comment (juriczech) [reply suite-native/graph/src/reducers/types.ts]: What's the reason to have reducer in folder then? Because then it's just useless folder for 1 file that can be moved level down. So I think we should either move everything from this folder to src, or group them together here with different name? This would then be useless nesting of folders and creates inconsistencies. What do you think? Move reducer to root as well?

File: suite-native/state/src/configureStore.ts (hunk 3 above)
Comment (Nodonisko) [reply suite-native/state/src/configureStore.ts]: I don't see reason why we should change something that works well and it's according to docs. Also I think this is going to break thunks and possible some other stuff because configure store automatically adding some middlewares and enhancers. 

We can migrate it to manual way in case we will need to do something custom, but we don't need that now and probably won't ever need.

Comment (juriczech) [reply suite-native/state/src/configureStore.ts]: Yeah, I agree. Fixed already.

File: suite-native/graph/src/reducers/types.ts (hunk 2 above)
Comment (Nodonisko) [reply suite-native/graph/src/reducers/types.ts]: Yea, why not, keep it flat. We can create folder later if there will be more reducers.

Comment (juriczech) [reply suite-native/graph/src/reducers/types.ts]: https://github.com/trezor/trezor-suite/pull/5838/commits/718a455a86a047e0ce3636bde0cd4e79b3a96fc7

File: suite-native/atoms/src/AssetItem.tsx [hunk 5]
```diff
@@ -16,6 +16,11 @@ type AssetItemProps = {
     iconName: CryptoIconName;
 };
 
+export const assetColorConfig: Record<string, string> = {
```
Comment (Nodonisko) [inline suite-native/atoms/src/AssetItem.tsx:20]: I would go with simple `as const` and you can remove whole Record<string, string> thing.

File: suite-native/graph/src/appGraphReducer.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
+import { createSlice, PayloadAction } from '@reduxjs/toolkit';
```
Comment (Nodonisko) [inline suite-native/graph/src/appGraphReducer.ts]: prefix it with app doesn't make sense, but anyway I would suggest just naming it reducer (same logic as for types.ts) and moving it to package root

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-native/graph/jest.config.js [hunk 7]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+    preset: '../../jest.config.base.js',
```
Comment (bouchja1) [inline suite-native/graph/jest.config.js]: Is this file necessary here? We are passing it with no tests for now.

File: suite-native/graph/src/components/Graph.tsx [hunk 8]
```diff
@@ -0,0 +1,47 @@
+import React from 'react';
+import { LineGraph } from 'react-native-graph';
+
+import { prepareNativeStyle, useNativeStyles } from '@trezor/styles';
+import { Box } from '@suite-native/atoms';
+
+import { TimeSwitch } from './TimeSwitch';
+
+const dummyData = [
```
Comment (Nodonisko) [inline suite-native/graph/src/components/Graph.tsx]: This should not be in this package, dummy data should be in place where this package is used, afterwards we will only change source of data

File: suite-native/graph/jest.config.js (hunk 7 above)
Comment (juriczech) [reply suite-native/graph/jest.config.js]: Guess not, but it's in nearly all packages in /suite-native so I am just following the convention we've estabilished. I think it was discussed before in your navigation PR? Not sure though. When tests get added (I guess quite soon since we're getting close to doing business logic) it will need to extend this config so I would leave it. If you think I should remove it, it should be removed from all other packages where there are not tests as well.

File: suite-native/graph/package.json [hunk 9]
```diff
@@ -0,0 +1,26 @@
+{
+    "name": "@suite-native/graph",
```
Comment (Nodonisko) [inline suite-native/graph/package.json]: I am not sure about name, I would probably suggest something more specific like `home-graph` because there is already reducer so it can be only used in one place, because it will share state everywhere.

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-native/graph/src/appGraphReducer.ts (hunk 6 above)
Comment (juriczech) [reply suite-native/graph/src/appGraphReducer.ts]: Okay, changed it also for settings https://github.com/trezor/trezor-suite/pull/5838/commits/527b44a9dad6f51456a6b2cbcf93fb7fef034c36

File: suite-native/atoms/src/AssetItem.tsx (hunk 5 above)
Comment (juriczech) [reply suite-native/atoms/src/AssetItem.tsx:20]: I tried this, but ran into error with index signature when passing the prop to progress bar (see type of `cryptoCurrencySymbol`)

File: suite-native/graph/jest.config.js (hunk 7 above)
Comment (Nodonisko) [reply suite-native/graph/jest.config.js]: I would prefer to remove tests stuff from packages where we don't have tests, but this package is definitely worth of having some tests because there is some logic (Time switching) that should be tested. So it would be nice to add them right now.

Comment (Nodonisko) [reply suite-native/graph/jest.config.js]: You can use this lib https://testing-library.com it should pretty easy with that to add some basic tests for Graph component.

File: suite-native/atoms/src/AssetItem.tsx (hunk 5 above)
Comment (dahaca) [reply suite-native/atoms/src/AssetItem.tsx:20]: I would also have this config somewhere in another file but for the MVP it's too nitpick-y, I suppose :)

File: suite-native/graph/package.json (hunk 9 above)
Comment (dahaca) [reply suite-native/graph/package.json]: Is a separate package for the graph is really needed? 🤔

File: suite-native/graph/src/components/TimeSwitch.tsx [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    flexDirection: 'row',
+    justifyContent: 'space-around',
+}));
+
+export const TimeSwitch = () => {
+    const { applyStyle } = useNativeStyles();
+
+    return (
+        <Box style={applyStyle(timeSwitchStyle)}>
+            {Object.values(timeSwitchItems).map(item => (
```
Comment (dahaca) [inline suite-native/graph/src/components/TimeSwitch.tsx]: Could be memoized?

File: suite-native/graph/src/components/TimeSwitchItem.tsx [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+const textStyle = prepareNativeStyle<TextStyleProps>((utils, { isSelected }) => ({
+    ...utils.typography.hint,
+    color: isSelected ? utils.colors.forest : utils.colors.gray500,
+    textTransform: 'uppercase',
+}));
+
+export const TimeSwitchItem = ({ value, shortcut }: TimeSwitchItemProps) => {
+    const dispatch = useDispatch();
+    const { applyStyle } = useNativeStyles();
+    const selectedTimeFrame = useSelector(getSelectedTimeFrame);
```
Comment (dahaca) [inline suite-native/graph/src/components/TimeSwitchItem.tsx]: nit: Would be cool to keep the order of hooks used at least a bit consistent across the components perhaps? For example, I would always put selectors and state above everything.

File: suite-native/graph/src/reducer.ts [hunk 12]
```diff
@@ -0,0 +1,30 @@
+import { createSlice, PayloadAction } from '@reduxjs/toolkit';
```
Comment (dahaca) [inline suite-native/graph/src/reducer.ts]: I've just noticed you are using the toolkit, awesomeeeee!!!!! 🥳🥳🥳🥳🥳

File: suite-native/graph/src/reducer.ts [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    name: 'appGraph',
+    initialState,
+    reducers: {
+        setSelectedTimeFrame: (state, { payload }: PayloadAction<TimeFrameValues>) => {
+            state.selectedTimeFrame = payload;
+        },
+    },
+});
+
+export const getSelectedTimeFrame = (state: RootState) => state.appGraph.selectedTimeFrame;
```
Comment (dahaca) [inline suite-native/graph/src/reducer.ts]: Perhaps we could keep selectors in a separate file?

File: suite-native/graph/src/reducer.ts [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+interface AppGraphState {
+    selectedTimeFrame: TimeFrameValues;
+}
+
+const initialState: AppGraphState = {
+    selectedTimeFrame: 'day',
+};
+
+interface RootState {
```
Comment (dahaca) [inline suite-native/graph/src/reducer.ts]: It should be in `store.ts`, right?

File: suite-native/module-home/src/components/PortfolioGraph.tsx [hunk 15]
```diff
@@ -1,20 +1,53 @@
...
+    return (
+        <Box>
+            <Text variant="titleLarge">$3,720</Text>
+            <Box flexDirection="row" alignItems="center">
+                <Box marginRight="small">
+                    <Text variant="hint" color="gray600">
+                        Today, 15:45
+                    </Text>
+                </Box>
+                <Box style={applyStyle(arrowStyle, { direction: 'down' })} />
```
Comment (dahaca) [inline suite-native/module-home/src/components/PortfolioGraph.tsx]: Will be an icon, as we discussed ;)

**Review by dahaca — CHANGES_REQUESTED** [review]

File: suite-native/graph/src/components/TimeSwitchItem.tsx (hunk 11 above)
Comment (juriczech) [reply suite-native/graph/src/components/TimeSwitchItem.tsx]: Yeah no problem. I like to use the "imports" (dispatch, styles, etc) on the top and then state and vars. But it makes more sense to have state on top I guess.

File: suite-native/atoms/src/AssetItem.tsx (hunk 5 above)
Comment (juriczech) [reply suite-native/atoms/src/AssetItem.tsx:20]: I thought about it as well. Problem is that this is in `/atoms` and if I put it in `/module-home`, it will create circular dependency. What about `constants.ts` in src of atoms? @Nodonisko @dahaca

File: suite-native/graph/src/reducer.ts (hunk 13 above)
Comment (bouchja1) [reply suite-native/graph/src/reducer.ts]: > Perhaps we could keep selectors in a separate file?

I think it is better to keep a related logic together (when using the Redux Toolkit `createSlice` API) to make it easier to find and update that code. It is similar way as the ["ducks" pattern](https://github.com/erikras/ducks-modular-redux). But perhaps we should not call the file _reducer_ because, in fact, it is a _slice_ with all its state, actions, selectors, exporting the reducer...

The original duck approach would be:

- `/src`
  - `/features`
    - `/graph`
      - `TimeSwitchItem.tsx`: a React component that shows the UI for the selected time frame
      - `graphSlice.js`: the Redux logic for the graph feature

I am not sure if this is fully applicable for our purposes so would keep just with renaming the file:

- `/src`
    - `graphSlice.js`

And in case of more reducers in a package:

- `/src`
  - `/features`
      - `someSlice.js`
      - `someOtherDomainSlice.js`

Comment (juriczech) [reply suite-native/graph/src/reducer.ts]: Yes, I agree that all the redux stuff can stay in one file. Good point with the filename as well, I will rename it to slice. 

One more benefit to what @bouchja1 just mentioned is that once that file grows too large, often it's a good indicator about splitting that slice into 2. Slices should be semantically partitioned into small chunks of feature related logic. Some slices might even be just CRUD operations (using the `entityAdapter` from redux-toolkit with some UI logic as well maybe). But once there is more than 20 properties, usually it can be divided into more slices.

File: suite-native/graph/src/reducer.ts (hunk 14 above)
Comment (juriczech) [reply suite-native/graph/src/reducer.ts]: This is `RootState` of this slice. Bad name I guess. I will rename it to `SliceState`

File: suite-native/module-home/src/components/PortfolioGraph.tsx (hunk 15 above)
Comment (juriczech) [reply suite-native/module-home/src/components/PortfolioGraph.tsx]: https://github.com/trezor/trezor-suite/pull/5838/commits/8d03ac13bbdc855f35c585b400951d9764b552cb

File: suite-native/graph/src/reducer.ts (hunk 14 above)
Comment (juriczech) [reply suite-native/graph/src/reducer.ts]: https://github.com/trezor/trezor-suite/pull/5838/commits/8d03ac13bbdc855f35c585b400951d9764b552cb

File: suite-native/graph/src/reducer.ts (hunk 13 above)
Comment (juriczech) [reply suite-native/graph/src/reducer.ts]: Renamed to `slice.ts` https://github.com/trezor/trezor-suite/pull/5838/commits/8d03ac13bbdc855f35c585b400951d9764b552cb

File: suite-native/graph/src/components/TimeSwitchItem.tsx (hunk 11 above)
Comment (juriczech) [reply suite-native/graph/src/components/TimeSwitchItem.tsx]: https://github.com/trezor/trezor-suite/pull/5838/commits/8d03ac13bbdc855f35c585b400951d9764b552cb

File: suite-native/graph/src/components/TimeSwitch.tsx (hunk 10 above)
Comment (juriczech) [reply suite-native/graph/src/components/TimeSwitch.tsx]: Good point :)  https://github.com/trezor/trezor-suite/pull/5838/commits/8d03ac13bbdc855f35c585b400951d9764b552cb

File: suite-native/graph/src/components/Graph.tsx (hunk 8 above)
Comment (juriczech) [reply suite-native/graph/src/components/Graph.tsx]: https://github.com/trezor/trezor-suite/pull/5838/commits/8d03ac13bbdc855f35c585b400951d9764b552cb

File: suite-native/graph/src/components/TimeSwitch.tsx (hunk 10 above)
Comment (Nodonisko) [reply suite-native/graph/src/components/TimeSwitch.tsx]: Please don't do that, cost of memoization is probably higher that cost of renredering. This component should be nearly never rerendered because it has no props, no internal or redux state... even if yes, cost rerendering that few items here will be very very small.

I am not saying memoization can be actually faster, but there is golden rule for doing optimalizations and that rule is: **Don't do optimalizations without proper measurement, because you can make things worse than it was before. No optimilazation is better than bad and blindly done optimalization.**

File: suite-native/graph/src/reducer.ts (hunk 13 above)
Comment (Nodonisko) [reply suite-native/graph/src/reducer.ts]: I like what you guys said about keeping things it together. Context is important and it's even more important to not break it.

I kind of dislike filename `slice`, I agree that `reducer` is also not best, but I need to think about it.

File: suite-native/atoms/src/AssetItem.tsx (hunk 5 above)
Comment (Nodonisko) [reply suite-native/atoms/src/AssetItem.tsx:20]: When circular dependecy occurs it's strong signal that you need new package that will be used in both places. We will probably need some kind of suite-native config/constants package, but I think we should not do that in this PR. Let's only add TODO comment here, that this needs to be solved.

File: suite-native/graph/package.json (hunk 9 above)
Comment (Nodonisko) [reply suite-native/graph/package.json]: It's kind of standalone functionality with lot of logic/components around that has potential to grow even more. 

If I place this to the `module-home` package for example, it can work too, but it not that future proof. Over the time `module-home` will grow and could change quite often. With every change you need to rerun lint, type-check, tests for whole `module-home`. With graph as package, you are reducing amount of these checks needed to run, it's partially explained here https://www.notion.so/satoshilabs/Code-improvements-9-breaking-the-monolith-528bf80c34ba4a6bba3f335e60ff29e8#402adaee3192443d8ce477a970207d63

There is also not-a-low potential that we will need to use graph in more that one place and it's much easier to just little bit tweak this package to be reusable then cut it out-of much bigger package.

File: suite-native/atoms/src/AssetItem.tsx (hunk 5 above)
Comment (juriczech) [reply suite-native/atoms/src/AssetItem.tsx:20]: Yep, will do.

File: suite-native/graph/jest.config.js (hunk 7 above)
Comment (juriczech) [reply suite-native/graph/jest.config.js]: Will be done in separate PR.

File: suite-native/graph/src/components/TimeSwitch.tsx (hunk 10 above)
Comment (dahaca) [reply suite-native/graph/src/components/TimeSwitch.tsx]: You are right, it's just when there is a `.map()` in the render memo usually should save some renders. That component is not big and doesn't even have props, yes. I

File: suite-native/graph/src/reducer.ts (hunk 13 above)
Comment (dahaca) [reply suite-native/graph/src/reducer.ts]: Honestly, I would have a `state` folder in each package with related redux logic and have `reducer`, `selectors` and `actions`. I don't like having selectors in `reducers` :( Yes, simple actions are exported from the reducer, that is true, the `actions` file would be for thunks.

Comment (juriczech) [reply suite-native/graph/src/reducer.ts]: I am still going to say that thunks can be in slice as well. We actually had this format in my last job where we used redux-toolkit and it worked quite well. I would at least give it a try and we can always refactor? :) Slice is fine by me + another file for types / constants and it works pretty well imho. Of course, we can discuss it further in code improvements :)

Comment (Nodonisko) [reply suite-native/graph/src/reducer.ts]: Actions, reducer, selectors are very closely tied and one can't exist without another one. Placing them to separate files only results to constantly switching context between these files.

Nearly always you are doing some changes in action you need to that change to reducer and vice versa. What are benefits of having multiple files that would overweight disadvantage of force people to go to two or three different files to change one thing? Or even more when you have constants file for action types in one more extra file like we do in suite.

Same goes for the the thunks, thunk is usually some code that is also in some way dependent on state of slice, selector or need to dispatch another action from thunks.

I am pretty sure that with reduced boilerplate thanks to toolkit, most of these slices files be way under 100 lines of code which could be really easily handled in one file.

Comment (bouchja1) [reply suite-native/graph/src/reducer.ts]: I am still for putting things together into a single slice and would say it is about getting used to a new approach with redux-toolkit. I worked with "features" approach before and it really simplifies usage. Even Mark Erickson - Redux Maintainer - [recommends](https://www.reddit.com/user/acemarke/) a single slice file... there are also links ([[1](https://redux.js.org/style-guide/#structure-files-as-feature-folders-with-single-file-logic)], [[2](https://redux.js.org/tutorials/essentials/part-2-app-structure)]) to redux docs pointing to a single file logic as the best approach.

File: suite-native/home-graph/src/components/TimeSwitch.tsx [hunk 16]
```diff
@@ -38,7 +38,7 @@ const timeSwitchStyle = prepareNativeStyle(() => ({
     justifyContent: 'space-around',
 }));
 
-export const TimeSwitch = () => {
+export const TimeSwitch = memo(() => {
```
Comment (dahaca) [inline suite-native/home-graph/src/components/TimeSwitch.tsx]: By the way, I meant memoizing the map:

```
const switchItems = useMemo(() => Object.values(timeSwitchItems).map(item => (
   <TimeSwitchItem key={item.value} shortcut={item.shortcut} value={item.value} />
))}, [])
```

**Review by dahaca — APPROVED** [review]
Since the redux structure discussion is still ongoing, let's not solve it in this PR. Good job!

File: suite-native/graph/src/reducer.ts (hunk 13 above)
Comment (dahaca) [reply suite-native/graph/src/reducer.ts]: @bouchja1 I have read the recommendations and actually I do agree! The only difference is that we have a bundle per domain / page already, so it's not about putting in different files and grouping by type _(selector / reducer, etc)_. Since the Redux code is already located next to the place where it is going to be used, it is kinda like a duck. So separating into different files or not is mostly about "do we want one large file to store the code, or several smaller files and why". I don't think switching would be be a problem, while related, selectors are not always changing together with the state. You might need a new selector without changing the shape of state. Same with thunks. You might need to create a new thunk for a new feature that modifies existing state using existing action creators. Yes, it is not always the case! At times you would have to modify the reducer and then add a selector in a separate file. But imagine having a large reducer with complex thunks and and enormous switch in the middleware. I would not want to scroll 700 lines each time to reach the thunk/selector/reducer/middleware I wanna reach :D

File: suite-native/home-graph/src/components/TimeSwitch.tsx (hunk 16 above)
Comment (Nodonisko) [reply suite-native/home-graph/src/components/TimeSwitch.tsx]: I am pretty sure this won't have any effect in terms of performance for such a few items. But who knows, if you really want to do it, at least don't shoot into dark and if you really one to do it, them do please measure it. Could be measurement of rerenders and also speed of render.

File: suite-native/graph/src/reducer.ts (hunk 13 above)
Comment (Nodonisko) [reply suite-native/graph/src/reducer.ts]: One domain actually doesn't mean just one reducer, sometimes you will have multiple state slices in one package. Then you will have to have multiple folders with multiple reducers/actions/selectors files and that is lot file system noise and in the end sum of lines in one of these folders don't have over 150 lines of code. 

Files are usually much smaller than 700 lines, reducers usually have like 40 lines of code and even less for selectors. Only thunks could be quite long and you can easily solve this placing reducers and selector at the of file, so you can easily navigate in file.

**Review by Nodonisko — APPROVED** [review]

**Comment by juriczech** [comment]
Tests failing on unrelated packages. Merging.
